docs/unit-testing.md

```diff
@@ -0,0 +1,211 @@
+        res := gobatcher.NewAzureSharedResource("accountName", "containerName", 10000).
+            WithFactor(1000).
+            WithMocks(container, blob)
+        var wg sync.WaitGroup
+        wg.Add(1)
+        res.AddListener(func(event string, val int, msg string, metadata interface{}) {
+            switch event {
+            case gobatcher.ErrorEvent:
+                panic(errors.New(msg))
+            case gobatcher.ProvisionDoneEvent:
+                wg.Done()
+            }
+        })
+        err := res.Start(context.Background())
+        assert.NoError(t, err)
+        wg.Wait() // wait for provisioning (which is an async process) to finish
+        container.AssertExpectations(t)
+        blob.AssertExpectations(t)
+    }
+    ```
+
+### ProvisionedResource
+
+The ProvisionedResource does not need a `WithMocks()` method as it does not provision any resources in Azure. You can mock this interface as you would any other interface.
+
+### RateLimiter
+
+The RateLimiter interface allows you to create your own RateLimiters and use them with Batcher. However, this is outside of the scope of this unit test document.
+
+## Using events
+
+Both Batcher and RateLimiters raise events that you can interrogate in your unit tests to validate expected behaviors. Consider the following implementation:
+
+```go
+func WriteString(batcher gobatcher.IBatcher, data []string) {
+    var wg sync.WaitGroup
+    wg.Add(len(data))
+    watcher := gobatcher.NewWatcher(func(batch []gobatcher.IOperation) {
+        fmt.Println("START-OF-BATCH")
+        for _, op := range batch {
+            val := op.Payload().(string)
+            fmt.Println(val)
+        }
+        fmt.Println("END-OF-BATCH")
+        for i := 0; i < len(batch); i++ {
+            wg.Done()
+        }
+    })
+    for i, element := range data {
+        cost := 100 + i // do some fancy cost calculation
+        if err := batcher.Enqueue(gobatcher.NewOperation(watcher, uint32(cost), element, true)); err != nil {
+            panic(err)
+        }
+    }
+    wg.Wait()
+}
+```
+
+Then consider the following unit test that validates that all operation costs were 100 or greater and that each operation is put into a separate batch due to that cost:
+
+```go
+func TestWriteString_CostIs100OrMore(t *testing.T) {
+    res := gobatcher.NewProvisionedResource(1000)
+    // NOTE: The FlushInterval is 100ms so there will be 10 flushes per second with 100 capacity each, so operations that are 100 or more should be in their own batches
+    batcher := gobatcher.NewBatcher().
+        WithRateLimiter(res).
+        WithFlushInterval(100 * time.Millisecond).
+        WithEmitBatch()
+    var batches uint32 = 0
+    batcher.AddListener(func(event string, val int, msg string, metadata interface{}) {
+        switch event {
+        case gobatcher.BatchEvent:
+            assert.Equal(t, 1, val)
+            atomic.AddUint32(&batches, 1)
+            batch := metadata.([]gobatcher.IOperation)
```

Comment:
```suggestion
            batch := metadata.([]gobatcher.Operation)
```

Reply:
good catch. Those were written before we made the interface change. They are fixed now.